### ISSUE TITLE
Add `SessionConfig` reference to `ScalarFunctionArgs`

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#13519]
### Is your feature request related to a problem or challenge?

- Now that https://github.com/apache/datafusion/pull/13290 is merged (thanks @joseph-isaacs) we have a place to add new 

@Omega359  [noted](https://github.com/apache/datafusion/pull/13290#discussion_r1852255032) that by adding `SessionConfig` to the `ScalarFunctionArgs`   unblock several tasks such as
- https://github.com/apache/datafusion/issues/13212
- https://github.com/apache/datafusion/issues/12892
- https://github.com/apache/datafusion/issues/10744

### Describe the solution you'd like

Add `&SessionConfig` to `ScalarFunctionArgs`, and ideally add a test that shows the config gets through.

### Additional context

We should try and get this done before DataFusion 44 is released so it isn't a breaking change
- #13334

**Comment by Omega359** [comment]
Well, small issue. `ScalarFunctionArgs` is in `datafusion-expr` which can't use `SessionConfig` directly since it's in `datafusion-execution` which depends on `datafusion-expr` which would cause a circular dependency. We can use `ConfigOptions` there which would give us access to all config except the opaque extensions in `SessionConfig` which I think is acceptable.

**Comment by niebayes** [comment]
Great work!

**Comment by Omega359** [comment]
Started delving into this trying to find a good way to impl. Trying to add `config_options: ConfigOptions` to `ScalarFunctionExpr` Having lots of fun with eq and hash with SessionConfig and f64 atm.

**Comment by alamb** [comment]
I wonder if we can use this trait: https://docs.rs/datafusion/latest/datafusion/catalog/trait.Session.html

**Comment by Omega359** [comment]
We could but it wouldn't solve the issue. The issue is PhysicalExpr requires implementations to impl Eq and Hash or to have implementations for DynEq and DynHash. That is fine until something like SessionConfig which doesn't is introduced. I am looking at updating the either the 'config_namespace' macro or add explicit implementations for 'ScalarFunctionExpr' to try and impl either of the above and use f64.to_bits() and f64::from_bits(..) to handle the problematic f64's in the config. I think it's possible

**Comment by jayzhan211** [comment]
I think it makes sense to me to move `SessionConfig` to common crate

**Comment by Omega359** [comment]
take

**Comment by alamb** [comment]
> It makes sense to me to move `SessionConfig` to common or common-runtime crate

I agree

**Comment by Omega359** [comment]
I mistyped above - I meant we couldn't use SessionContext, not SessionConfig. Sorry about the confusion.

**Comment by jayzhan211** [comment]
I guess we can also add `nullable` info to `ScalarFunctionArgs` https://github.com/apache/datafusion/issues/11923

**Comment by alamb** [comment]
> I guess we can also add `nullable` info to `ScalarFunctionArgs` #11923

It might already be present in `ScalarFunctionArgs::data_type` 🤔 : https://github.com/apache/datafusion/blob/f2de2c4cc2009d9b6965f7951fd543e1974fcd2c/datafusion/expr/src/udf.rs#L336

**Comment by jayzhan211** [comment]
Not really, `DataType` has no nullable info, we have to send `nullable` to `ScalarFunctionArgs`

https://github.com/apache/datafusion/blob/6c9355d5be8b6045865fed67cb6d028b2dfc2e06/datafusion/physical-expr/src/scalar_function.rs#L148-L153

```rust
        // evaluate the function
        let nullable = self.nullable;
        let output = self.fun.invoke_with_args(ScalarFunctionArgs {
            args: inputs.as_slice(),
            number_rows: batch.num_rows(),
            return_type: &self.return_type,
            nullable,
        })?; 
```

**Comment by findepi** [comment]
> Add SessionConfig reference to ScalarFunctionArgs

This is pretty heavy dependency.
SessionConfig contains a LOT of information no scalar function should depend on and very little that has legitimate usage.
I would prefer not adding SessionConfig to ScalarFunctionArgs. Rather, we could add a stripped down config object dedicated to scalar functions. If time zone is all we need, we should add the time zone directly to `ScalarFunctionArgs` (like proposed in https://github.com/apache/datafusion/pull/16573).

**Comment by alamb** [comment]
https://docs.rs/datafusion/latest/datafusion/execution/config/struct.SessionConfig.html is mostly a wrapper around the ConfigOptions

If we just want specific fields, we could perhaps reuse the existing ExecutionProps

I think the challenge is that 
1. The subset of `ConfigOptions` that functions might want to access is not clear to me
2. User defined functions may well want to use settings we haven't anticipated (and SessionConfig has a user defined way to pass it)

That doesn't mean we couldn't copy a subset of the fields into ScalarFunctionArgs, it just seems unecessary to try and pick that subset 🤔

**Comment by findepi** [comment]
> If we just want specific fields, we could perhaps reuse the existing ExecutionProps

Are AliasGenerator and VarProviders relevant to function implementors?

> The subset of ConfigOptions that functions might want to access is not clear to me

It's not clear to me either.  We know about session time zone. Session start time may be relevant.
Do we need to know all of them from the start, or just anticipate we gonna evolve set of what's accessible as the need arises?

By avoiding full SessionConfig / ConfigOptions, I would like to retain the state where it's very easy to invoke a function, either from a test, a benchmark, or any other execution context. 

Exposing sub-configs such SqlParserOptions, ExplainOptions, or OptimizerOptions to UDFs looks like breaking abstractions, or invitation to break abstractions.

**Comment by Omega359** [comment]
I think my next attempt at solving this was to take the same approach that async-udf took - which is to move the invocation of the udf into an [ExecutionPlan](https://github.com/apache/datafusion/blob/0143b20de3f9e25ab8d87ef4b96b78c28e462bee/datafusion/physical-plan/src/async_func.rs#L130). That approach provides the full config_options.

**Comment by findepi** [comment]
> Exposing sub-configs such SqlParserOptions, ExplainOptions, or OptimizerOptions to UDFs looks like breaking abstractions, or invitation to break abstractions.


In light of a push to reduce breaking changes like https://github.com/apache/datafusion/issues/16622 (and https://github.com/apache/datafusion/pull/16078, https://github.com/apache/datafusion/pull/16541, https://github.com/apache/datafusion/issues/13648) we could try to be more judicious about growing the public API. IF we don't need to expose full SessionConfig / ConfigOptions in ScalarFunctionArgs let's maybe not expose them.

**Comment by alamb** [comment]
> In light of a push to reduce breaking changes like https://github.com/apache/datafusion/issues/16622 (and https://github.com/apache/datafusion/pull/16078, https://github.com/apache/datafusion/pull/16541, https://github.com/apache/datafusion/issues/13648) we could try to be more judicious about growing the public API. IF we don't need to expose full SessionConfig / ConfigOptions in ScalarFunctionArgs let's maybe not expose them.

My counter argument is that having to add new fields to ExecutionProps basically increases the public API over time. `ConfigOptions` is already public so it is my opinion that exposing just the `ConfigOptions` is less API surface area. I can see how there are alternate opinions to this

Here is what passing in the entire ConfigOptions looks like (it is pretty straightforward actually): https://github.com/apache/datafusion/pull/16661

**Comment by findepi** [comment]
BTW i solved my problem by adding the needed fields into the UDF struct during it construction and then I don't need any more info at simplify() & invoke() time. When doing so, I had this observation -- https://github.com/apache/datafusion/issues/16677

**Comment by alamb** [comment]
Closed in https://github.com/apache/datafusion/pull/16970
